### PR TITLE
Migrate transactions

### PR DESCRIPTION
**Pull request by schoren**
This PR...

## Changes

-

## Fixes

-

## Checklist

- [ ] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

File: server/executor/transaction_runner_test.go [hunk 1]
```diff
@@ -88,34 +91,25 @@ func TestTransactionRunner(t *testing.T) {
...
-	return db, rawDB, clean
+	return db, rawDB
 }
 
-func runTransactionRunnerTest(t *testing.T, withErrors bool, assert func(t *testing.T, actual model.TransactionRun)) {
+func runTransactionRunnerTest(t *testing.T, withErrors bool, assert func(t *testing.T, actual tests.TransactionRun)) {
 	ctx := context.Background()
-	rawDB := testmock.GetRawTestingDatabase()
-	db, rawDB, clear := getDB()
-	defer clear()
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/executor/transaction_runner_test.go:118]: each test uses a new, clean DB, so no need to close it

File: server/model/test_json.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	jt.SpecsJSON = make([]jsonSpec, 0, t.Specs.Len())
+	t.Specs.ForEach(func(key SpanQuery, val NamedAssertions) error {
+		// assertions := make([]string, len(val.Assertions))
+		// for i, a := range val.Assertions {
+		// 	assertions[i] = string(a)
+		// }
+
+		jt.SpecsJSON = append(jt.SpecsJSON, jsonSpec{
+			Name: val.Name,
+			// Assertions: assertions,
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/model/test_json.go:52]: actually to delete it :P

File: server/model/test_json_test.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			}),
+	}
+
+	encoded, err := json.Marshal(original)
+	require.NoError(t, err)
+	fmt.Println(string(encoded))
+
+	decoded := model.Test{}
+	err = json.Unmarshal(encoded, &decoded)
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/model/test_json_test.go:52]: This doesn't change how things are persisted in DB. That is covered by other tests, including e2e, tracetesting and units.

File: server/model/tests.go [hunk 4]
```diff
@@ -34,23 +35,23 @@ type (
...
 	}
 
 	TriggerType string
 
 	Trigger struct {
-		Type    TriggerType
-		HTTP    *HTTPRequest
-		GRPC    *GRPCRequest
-		TRACEID *TRACEIDRequest
+		Type    TriggerType     `json:"triggerType"`
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/model/tests.go:51]: I might be confused, but I'm almost sure this is called `triggerType` in the [openapi specs](https://github.com/kubeshop/tracetest/blob/10b95850233b8f97928cab69bcfc9b4271745047/api/triggers.yaml#L7)

File: server/model/tests.go [hunk 5]
```diff
@@ -34,23 +35,23 @@ type (
...
+		HTTP    *HTTPRequest    `json:"http,omitempty"`
+		GRPC    *GRPCRequest    `json:"grpc,omitempty"`
+		TraceID *TRACEIDRequest `json:"traceid,omitempty"`
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/model/tests.go:54]: it was always in lowercase, I just change the Go name, not the JSON name

File: server/pkg/sqlutil/search.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+	}
+
+	sql += condition
+	params = append(params, CleanSearchQuery(query))
+
+	return sql, params
+}
+
+func CleanSearchQuery(query string) string {
+	return "%" + strings.ReplaceAll(query, " ", "%") + "%"
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/pkg/sqlutil/search.go:17]: it's a pseudo full-text search. If you have `search this` the search string is `%search%this%`. `%` is the wildcard for searches in pg, and also the preexisting behaviour

File: server/testdb/postgres_test.go [hunk 7]
```diff
@@ -48,28 +49,28 @@ func createTest(t *testing.T, db model.Repository) model.Test {
 	return createTestWithName(t, db, "first test")
 }
 
-func createTransaction(t *testing.T, db model.Repository) model.Transaction {
+func createTransaction(t *testing.T, tr *tests.TransactionsRepository) tests.Transaction {
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/testdb/postgres_test.go:52]: it was actually left over, good catch

File: server/tests/transactions_test.go [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+				}
+
+				// TODO: reenable this tests when we figure out how to test it
+				// problems:
+				//   1. sort fields do not map 1:1 with the actual items. Example "last_run" maps to item.summary.last_run.time
+				//   2. even if we pass some cusotm function to map fields, non augmented versions do not include the fields so we cannot assert them
+				// case rmtests.OperationListSortSuccess:
+				// 	transactionRepo.Create(context.TODO(), sample)
+				// 	transactionRepo.Create(context.TODO(), sample2)
+				// 	transactionRepo.Create(context.TODO(), sample3)
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/tests/transactions_test.go:216]: we could add an operation like `OperationAugmentedListSortedSuccess` or similar, but still doesn't the `fields do not map 1:1 to the actual items`. We need to figure that out, but I think we can advance this PR without this

File: server/tests/transaction_runs.go [hunk 9]
```diff
@@ -277,65 +273,65 @@ SELECT
...
-	run.Steps, err = td.getTransactionRunSteps(ctx, run)
+	run, err := td.readRunRow(stmt.QueryRowContext(ctx, runID, transactionID))
 	if err != nil {
-		return model.TransactionRun{}, err
+		return TransactionRun{}, err
 	}
+	// run.Steps, err = td.getTransactionRunSteps(ctx, run)
+	// if err != nil {
+	// 	return TransactionRun{}, err
+	// }
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/tests/transaction_runs.go:289]: no, this is supposed to be working

File: server/tests/transactions_test.go [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+	// }
+
+	// sample3 := tests.Transaction{
+	// 	ID:          "sample3",
+	// 	Name:        "Some Transaction",
+	// 	Description: "Do important stuff",
+	// 	StepIDs: []id.ID{
+	// 		"ezMn7bE4g",
+	// 	},
+	// }
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/tests/transactions_test.go:150]: no, those are only used for testing the list sort. The basic `OperationListSuccess` actually expects only one item

File: server/model/json_test.go [hunk 11]
```diff
@@ -148,51 +148,6 @@ func TestRunEncoding(t *testing.T) {
...
-
-	specs := maps.Ordered[model.SpanQuery, []model.Assertion]{}
-	specs = specs.MustAdd(model.SpanQuery(`span[tracetest.span.type = "http"]`), []model.Assertion{
-		model.Assertion(`attr:http.status = 200`),
-	})
-	oldTest := OldTest{
-		ID:               id.NewRandGenerator().ID(),
-		CreatedAt:        time.Now(),
-		Name:             "my test name",
-		Description:      "this is an old test using the old test format from version <= 0.7.2",
```
[the remark this answers is not in the document]
Comment (schoren) [reply server/model/json_test.go]: Forgot about this test, and it was actually regressing. Good catch!

File: server/http/controller.go [hunk 12]
```diff
@@ -511,6 +531,117 @@ func (c *controller) ExecuteDefinition(ctx context.Context, testDefinition opena
...
+			resp, err := c.doUpdateTransaction(ctx, transaction.ID, transaction)
+			if err != nil {
+				return resp, err
+			}
+		} else {
+			return resp, err
+		}
+	} else {
+		transactionID = id.ID(resp.Body.(openapi.Transaction).Id)
+	}
```
Comment (danielbdias) [inline server/http/controller.go:549]: We are not using the `transactionID` at the method start, so we could set it only after error handling
```suggestion
	// create or update transaction
	resp, err := c.doCreateTransaction(ctx, transaction)
	if err != nil {
		if errors.Is(err, errTransactionExists) {
			resp, err := c.doUpdateTransaction(ctx, transaction.ID, transaction)
			if err != nil {
				return resp, err
			}
		} else {
			return resp, err
		}
	} 
    
    transactionID := id.ID(resp.Body.(openapi.Transaction).Id)
```

File: server/model/json_test.go [hunk 13]
```diff
@@ -185,10 +185,12 @@ func TestOldAssertionSpecsFormatWithoutNames(t *testing.T) {
 
 	oldTestJson, err := json.Marshal(oldTest)
 	require.NoError(t, err)
+	t.Log(string(oldTestJson))
 
 	var test model.Test
 	err = json.Unmarshal(oldTestJson, &test)
 	require.NoError(t, err)
+	// spew.Dump(test)
```
Comment (danielbdias) [inline server/model/json_test.go:193]: Can we remove this comment?

File: server/pkg/sqlutil/search.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielbdias) [reply server/pkg/sqlutil/search.go:17]: :heart:

File: server/tests/transaction_runs.go [hunk 14]
```diff
@@ -57,6 +58,17 @@ INSERT INTO transaction_runs (
 )
 RETURNING "id"`
 
+const (
+	createSequeceQuery = `CREATE SEQUENCE IF NOT EXISTS "` + runSequenceName + `";`
+	dropSequeceQuery   = `DROP SEQUENCE IF EXISTS "` + runSequenceName + `";`
```
Comment (danielbdias) [inline server/tests/transaction_runs.go:63]:
```suggestion
	createSequenceQuery = `CREATE SEQUENCE IF NOT EXISTS "` + runSequenceName + `";`
	dropSequenceQuery   = `DROP SEQUENCE IF EXISTS "` + runSequenceName + `";`
```

**Review by danielbdias — APPROVED** [review]

**Pull request merged**